collections/src/main/scala/strawman/collection/immutable/List.scala

```diff
@@ -345,20 +345,76 @@ sealed trait List[+A]
   protected final def writeReplace(): AnyRef = new List.SerializationProxy(this)
 
   override def className = "List"
+
+  /** Builds a new list by applying a function to all elements of this list.
+    *  Like `xs map f`, but returns `xs` unchanged if function
+    *  `f` maps all elements to themselves (as determined by `eq`).
+    *
+    *  @param f      the function to apply to each element.
+    *  @tparam B     the element type of the returned collection.
+    *  @return       a list resulting from applying the given function
+    *                `f` to each element of this list and collecting the results.
+    *
+    *  @usecase def mapConserve(f: A => A): List[A]
+    *    @inheritdoc
+    */
+  @`inline` final def mapConserve[B >: A <: AnyRef](f: A => B): List[B] = {
+    // Note to developers: there exists a duplication between this function and `reflect.internal.util.Collections#map2Conserve`.
+    // If any successful optimization attempts or other changes are made, please rehash them there too.
+    //@tailrec
+    def loop(mappedHead: List[B] = Nil, mappedLast: ::[B], unchanged: List[A], pending: List[A]): List[B] = {
+      if (pending.isEmpty) {
+        if (mappedHead eq null) unchanged
+        else {
+          mappedLast.next = (unchanged: List[B])
```

Comment:
This did not compile on Dotty without the explicit widening. I get a poor error message without a position:
```
[info] Compiling 76 Scala sources and 5 Java sources to /Users/szeiger/code/scala/collection-strawman/collections/.jvm/target/scala-0.3/classes...
[error]
[error] covariant type A occurs in contravariant position in type strawman.collection.immutable.List[A] of value x$0
[error] one error found
```